extensions/aws/processors/FetchS3Object.cpp

```diff
@@ -0,0 +1,122 @@
+/**
+ * @file FetchS3Object.cpp
+ * FetchS3Object class implementation
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "FetchS3Object.h"
+
+#include <set>
+#include <memory>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace aws {
+namespace processors {
+
+const core::Property FetchS3Object::Version(
+  core::PropertyBuilder::createProperty("Version")
+    ->withDescription("The Version of the Object to download")
+    ->supportsExpressionLanguage(true)
+    ->build());
+const core::Property FetchS3Object::RequesterPays(
+  core::PropertyBuilder::createProperty("Requester Pays")
+    ->isRequired(true)
+    ->withDefaultValue<bool>(false)
+    ->withDescription("If true, indicates that the requester consents to pay any charges associated with retrieving "
+                      "objects from the S3 bucket. This sets the 'x-amz-request-payer' header to 'requester'.")
+    ->build());
+
+const core::Relationship FetchS3Object::Success("success", "FlowFiles are routed to success relationship");
+const core::Relationship FetchS3Object::Failure("failure", "FlowFiles are routed to failure relationship");
+
+void FetchS3Object::initialize() {
+  // Add new supported properties
+  updateSupportedProperties({Version, RequesterPays});
+  // Set the supported relationships
+  setSupportedRelationships({Failure, Success});
+}
+
+void FetchS3Object::onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &sessionFactory) {
+  S3Processor::onSchedule(context, sessionFactory);
+
+  context->getProperty(RequesterPays.getName(), get_object_params_.requester_pays);
+  logger_->log_debug("FetchS3Object: RequesterPays [%s]", get_object_params_.requester_pays ? "true" : "false");
+}
+
+bool FetchS3Object::getExpressionLanguageSupportedProperties(
+    const std::shared_ptr<core::ProcessContext> &context,
+    const std::shared_ptr<core::FlowFile> &flow_file) {
+  if (!S3Processor::getExpressionLanguageSupportedProperties(context, flow_file)) {
+    return false;
+  }
+  get_object_params_.bucket = std::move(bucket_);
+  get_object_params_.object_key = std::move(object_key_);
```

Comment:
it seems like `object_key_` and `bucket_` are a processor members, and we read/write them from possible multiple concurrent `onTrigger`s

Reply:
the same goes for, I think, `aws_credentials_provider_`and `s3_wrapper_` where we call `S3WrapperBase::setProxy` and `S3WrapperBase::setCredentials` from the `onTrigger`s

Reply:
Could you please elaborate on this a bit more? I am not that familiar with the scheduling of the onTrigger calls. Are you saying that it is possible that the same processor's onTrigger call can be scheduled on separate threads in the threadpool and depending on the timing they might overlap causing the concurrency?

Reply:
Yes, this can happen depending on configurations.
In most of the processors we just prevent this by having a mutex for the whole ontrigger call, in this case we might just protect some common resources.

Reply:
yes, as far as I know, `onSchedule` is guaranteed to be called once for a processor, but multiple `onTrigger`s can run concurrently on multiple threads, the number of such `onTrigger`s is controlled by the `max concurrent tasks` property of the processor

Reply:
of course we could always just ignore this property and override `getMaxConcurrentTasks` to return `1`, but I don't (yet) think there is an inherent reason to do so, for this processor (family)

Reply:
Thanks for the info and great catch! I removed the members that were modified in the onTrigger calls and added a mutex for the S3Wrapper calls to guarantee that the S3 configuration is not changed between AWS calls. Change is done in [e59580a](https://github.com/apache/nifi-minifi-cpp/pull/970/commits/e59580a5cbd1ef78a05da67664103fcfaeb4a062)